File: mobile/ios/iRadio/RadioAPI.swift
import Foundation

actor RadioAPI {
    private let baseURL: String
    private let session: URLSession

    var streamURL: String {
        return "\(baseURL)/stream/live"
    }

    var icecastBaseURL: String {
        return "\(baseURL)/stream"
    }

    init(baseURL: String) {
        self.baseURL = baseURL
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        self.session = URLSession(configuration: config)
    }

    func fetchConfig() async -> StationConfig? {
        guard let url = URL(string: "\(baseURL)/api/config") else { return nil }
        do {
            let (data, _) = try await session.data(from: url)
            return try JSONDecoder().decode(StationConfig.self, from: data)
        } catch {
            return nil
        }
    }

    func fetchNowPlaying() async -> NowPlayingData? {
        guard let url = URL(string: "\(baseURL)/api/now-playing") else { return nil }
        do {
            let (data, _) = try await session.data(from: url)
            return try JSONDecoder().decode(NowPlayingData.self, from: data)
        } catch {
            return nil
        }
    }

    func fetchListenerCount() async -> Int {
        guard let url = URL(string: "\(icecastBaseURL)/status-json.xsl") else { return 0 }
        do {
            let (data, _) = try await session.data(from: url)
            guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let icestats = json["icestats"] as? [String: Any] else { return 0 }

            if let source = icestats["source"] as? [String: Any] {
                return source["listeners"] as? Int ?? 0
            } else if let sources = icestats["source"] as? [[String: Any]] {
                let mount = sources.first { ($0["listenurl"] as? String)?.contains("/live") == true }
                return mount?["listeners"] as? Int ?? 0
            }
            return 0
        } catch {
            return 0
        }
    }

    func searchSong(title: String, artist: String? = nil) async -> SearchResult {
        var urlString = "\(baseURL)/api/search-song?title=\(title.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? title)"
        if let artist = artist, !artist.isEmpty {
            urlString += "&artist=\(artist.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? artist)"
        }
        guard let url = URL(string: urlString) else { return SearchResult(songs: nil, resolved: nil) }
        do {
            let (data, _) = try await session.data(from: url)
            return try JSONDecoder().decode(SearchResult.self, from: data)
        } catch {
            return SearchResult(songs: nil, resolved: nil)
        }
    }

    func submitRequest(body: SongRequestBody) async -> (Int, RequestResponse) {
        guard let url = URL(string: "\(baseURL)/api/request") else {
            return (0, RequestResponse(song: nil, error: "Invalid URL", suggestions: nil))
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        do {
            request.httpBody = try JSONEncoder().encode(body)
            let (data, response) = try await session.data(for: request)
            let httpResponse = response as? HTTPURLResponse
            let statusCode = httpResponse?.statusCode ?? 0
            let responseData = try JSONDecoder().decode(RequestResponse.self, from: data)
            return (statusCode, responseData)
        } catch {
            return (0, RequestResponse(song: nil, error: "Network error", suggestions: nil))
        }
    }
}

// MARK: - SSE Client

class SSEClient {
    private var task: URLSessionDataTask?
    private let url: URL
    private var onEvent: ((NowPlayingData) -> Void)?

    init(url: URL) {
        self.url = url
    }

    func connect(onEvent: @escaping (NowPlayingData) -> Void) {
        self.onEvent = onEvent
        var request = URLRequest(url: url)
        request.setValue("text/event-stream", forHTTPHeaderField: "Accept")
        request.timeoutInterval = TimeInterval.infinity

        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = TimeInterval.infinity
        config.timeoutIntervalForResource = TimeInterval.infinity
        let session = URLSession(configuration: config, delegate: SSEDelegate(handler: self), delegateQueue: nil)
        task = session.dataTask(with: request)
        task?.resume()
    }

    func disconnect() {
        task?.cancel()
        task = nil
    }

    fileprivate var buffer = ""

    fileprivate func processBuffer() {
        let lines = buffer.components(separatedBy: "\n")
        var eventType: String?
        var dataStr = ""

        for line in lines {
            if line.hasPrefix("event:") {
                eventType = line.replacingOccurrences(of: "event:", with: "").trimmingCharacters(in: .whitespaces)
            } else if line.hasPrefix("data:") {
                dataStr += line.replacingOccurrences(of: "data:", with: "").trimmingCharacters(in: .whitespaces)
            } else if line.isEmpty && !dataStr.isEmpty {
                if eventType == "now-playing", let data = dataStr.data(using: .utf8) {
                    if let nowPlaying = try? JSONDecoder().decode(NowPlayingData.self, from: data) {
                        DispatchQueue.main.async { [weak self] in
                            self?.onEvent?(nowPlaying)
                        }
                    }
                }
                eventType = nil
                dataStr = ""
            }
        }

        // Keep incomplete data in buffer
        if let lastNewline = buffer.lastIndex(of: "\n") {
            buffer = String(buffer[buffer.index(after: lastNewline)...])
        }
    }
}

private class SSEDelegate: NSObject, URLSessionDataDelegate {
    weak var handler: SSEClient?

    init(handler: SSEClient) {
        self.handler = handler
    }

    func urlSession(_ session: URLSession, dataTask: URLSessionDataTask, didReceive data: Data) {
        guard let str = String(data: data, encoding: .utf8) else { return }
        handler?.buffer += str
        handler?.processBuffer()
    }
}

File: builds/android/app/src/main/java/net/downstreamtech/rendezvox/data/RadioApi.kt
package net.downstreamtech.rendezvox.data

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import okhttp3.*
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

class RadioApi(private val baseUrl: String) {

    private val client = OkHttpClient.Builder()
        .connectTimeout(10, java.util.concurrent.TimeUnit.SECONDS)
        .readTimeout(30, java.util.concurrent.TimeUnit.SECONDS)
        .build()

    private val gson = Gson()

    val streamUrl: String
        get() = "$baseUrl/stream/live"

    val icecastBaseUrl: String
        get() = "$baseUrl/stream"

    suspend fun fetchConfig(): StationConfig = withContext(Dispatchers.IO) {
        val request = Request.Builder().url("$baseUrl/api/config")
            .cacheControl(CacheControl.FORCE_NETWORK).build()
        try {
            val response = client.newCall(request).await()
            val body = response.body?.string() ?: "{}"
            gson.fromJson(body, StationConfig::class.java)
        } catch (e: Exception) {
            StationConfig()
        }
    }

    suspend fun fetchNowPlaying(): NowPlayingData? = withContext(Dispatchers.IO) {
        val request = Request.Builder().url("$baseUrl/api/now-playing").build()
        try {
            val response = client.newCall(request).await()
            val body = response.body?.string() ?: return@withContext null
            gson.fromJson(body, NowPlayingData::class.java)
        } catch (e: Exception) {
            null
        }
    }

    suspend fun fetchListenerCount(): Int = withContext(Dispatchers.IO) {
        val request = Request.Builder().url("$icecastBaseUrl/status-json.xsl").build()
        try {
            val response = client.newCall(request).await()
            val body = response.body?.string() ?: return@withContext 0
            val json = gson.fromJson(body, Map::class.java)
            val icestats = json["icestats"] as? Map<*, *> ?: return@withContext 0
            val source = icestats["source"] ?: return@withContext 0

            when (source) {
                is Map<*, *> -> ((source["listeners"] as? Number)?.toInt() ?: 0)
                is List<*> -> {
                    val mount = source.filterIsInstance<Map<*, *>>()
                        .find { (it["listenurl"] as? String)?.contains("/live") == true }
                    (mount?.get("listeners") as? Number)?.toInt() ?: 0
                }
                else -> 0
            }
        } catch (e: Exception) {
            0
        }
    }

    suspend fun fetchRecentPlays(): List<RecentPlay> = withContext(Dispatchers.IO) {
        val request = Request.Builder().url("$baseUrl/api/recent-plays").build()
        try {
            val response = client.newCall(request).await()
            val body = response.body?.string() ?: return@withContext emptyList()
            val result = gson.fromJson(body, RecentPlaysResponse::class.java)
            result.plays
        } catch (_: Exception) {
            emptyList()
        }
    }

    suspend fun searchSong(title: String, artist: String? = null): SearchResult =
        withContext(Dispatchers.IO) {
            var url = "$baseUrl/api/search-song?title=${java.net.URLEncoder.encode(title, "UTF-8")}"
            if (!artist.isNullOrBlank()) {
                url += "&artist=${java.net.URLEncoder.encode(artist, "UTF-8")}"
            }
            val request = Request.Builder().url(url).build()
            try {
                val response = client.newCall(request).await()
                val body = response.body?.string() ?: return@withContext SearchResult()
                gson.fromJson(body, SearchResult::class.java)
            } catch (e: Exception) {
                SearchResult()
            }
        }

    suspend fun submitRequest(requestBody: RequestBody): Pair<Int, RequestResponse> =
        withContext(Dispatchers.IO) {
            val json = gson.toJson(requestBody)
            val body = json.toRequestBody("application/json".toMediaType())
            val request = Request.Builder()
                .url("$baseUrl/api/request")
                .post(body)
                .build()
            try {
                val response = client.newCall(request).await()
                val responseBody = response.body?.string() ?: "{}"
                val data = gson.fromJson(responseBody, RequestResponse::class.java)
                Pair(response.code, data)
            } catch (e: Exception) {
                Pair(0, RequestResponse(error = "Network error"))
            }
        }

    /**
     * Connect to the SSE endpoint for real-time now-playing updates.
     * Calls [onEvent] on each "now-playing" event with the parsed data.
     * Blocks the calling coroutine until cancelled.
     */
    suspend fun connectSSE(onEvent: (NowPlayingData) -> Unit): Unit =
        withContext(Dispatchers.IO) {
            val request = Request.Builder()
                .url("$baseUrl/api/sse/now-playing")
                .header("Accept", "text/event-stream")
                .build()

            try {
                val response = client.newCall(request).await()
                val reader = BufferedReader(InputStreamReader(response.body!!.byteStream()))
                var eventType: String? = null
                val dataBuilder = StringBuilder()

                while (true) {
                    val line = reader.readLine() ?: break

                    when {
                        line.startsWith("event:") -> {
                            eventType = line.removePrefix("event:").trim()
                        }
                        line.startsWith("data:") -> {
                            dataBuilder.append(line.removePrefix("data:").trim())
                        }
                        line.isEmpty() && dataBuilder.isNotEmpty() -> {
                            if (eventType == "now-playing") {
                                try {
                                    val data = gson.fromJson(
                                        dataBuilder.toString(),
                                        NowPlayingData::class.java
                                    )
                                    onEvent(data)
                                } catch (_: Exception) {}
                            }
                            eventType = null
                            dataBuilder.clear()
                        }
                    }
                }
            } catch (_: Exception) {
                // Connection lost — caller should reconnect
            }
        }

    private suspend fun Call.await(): Response = suspendCancellableCoroutine { cont ->
        enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                if (!cont.isCancelled) cont.resumeWithException(e)
            }
            override fun onResponse(call: Call, response: Response) {
                cont.resume(response)
            }
        })
        cont.invokeOnCancellation { cancel() }
    }
}
